### PR TITLE
provider/aws: Add resource_aws_volume_attachment

### PR DESCRIPTION
**Pull request by catsby**
Allows you to attach an existing EBS volume to an existing Instance.

**Usage:**

```javascript
provider "aws" {
  region="us-west-2"
}

resource "aws_instance" "web" {
  ami = "ami-21f78e11"
  availability_zone = "us-west-2a"
  instance_type = "t1.micro"
  tags {
    Name = "HelloWorld"
  }
}

resource "aws_ebs_volume" "example" {
  availability_zone = "us-west-2a" # must be same as Instance
  size = 5
}

resource "aws_volume_attachment" "ebs_att" {
  device_name = "/dev/sdh"
  volume_id = "${aws_ebs_volume.example.id}"
  instance_id = "${aws_instance.web.id}"
}

```

**Remaining:**
- [x] docs
- [x] acceptance test(s)

### CODE REVIEW

File: builtin/providers/aws/resource_aws_ebs_volume_test.go [hunk 1]
```diff
@@ -1,23 +1,59 @@
...
 				Config: testAccAwsEbsVolumeConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVolumeExists("aws_ebs_volume.test", &v),
+				),
 			},
 		},
 	})
 }
 
+func testAccCheckVolumeExists(n string, v *ec2.Volume) resource.TestCheckFunc {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_ebs_volume_test.go]: Upgraded this test while I was here, since I needed a `testAccCheckVolumeExists` method

Comment (phinze) [inline builtin/providers/aws/resource_aws_ebs_volume_test.go]: :chart_with_upwards_trend: :+1:

File: builtin/providers/aws/resource_aws_volume_attachment.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		return fmt.Errorf("Error attaching volume %s to instance %s: %s", vID, iID, err)
+	}
+
+	d.SetId(volumeAttachmentID(name, vID, iID))
+	return resourceAwsVolumeAttachmentRead(d, meta)
+}
+
+func attachVolumeFunc(conn *ec2.EC2, opts *ec2.AttachVolumeInput) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		va, err := conn.AttachVolume(opts)
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: I'm not used to `WaitForState` funcs actually performing the operation like this.

It seems weird that `VolumeIsInUse` is interpreted as "attaching" below - I'd expect the flow to be more like:

 * `AttachVolume` once
 * Handle any errors from that call.
 * _Then_ do a `StateRefreshFunc` that loops `DescribeVolumes` and waits for it to look right

**Comment by julienlavergne** [comment]
@phinze @catsby 

Great!

When do you think to merge this PR? ;) I can't wait for testing it!

**Comment by julienlavergne** [comment]
@phinze @catsby

Great!

When do you think to merge this PR? ;) I can't wait for testing it!

**Comment by catsby** [comment]
I plan to review the feedback provided, and maybe make an adjustment or two. I suspect we can merge this afternoon or tomorrow :smile:

**Comment by julienlavergne** [comment]
That’s wonderful !

great !

-- 
Julien Lavergne
Sent with Airmail

On May 26, 2015 at 4:37:17 PM, Clint (notifications@github.com) wrote:

I plan to review the feedback provided, and maybe make an adjustment or two. I suspect we can merge this afternoon or tomorrow

File: builtin/providers/aws/resource_aws_volume_attachment.go (hunk 2 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: Ah, so, this is a bit weird and maybe you can help me clarify things.
When you first call `AttachVolume`, it's possible that the volume you reference isn't totally ready. 

In TF's case, say that volume is another resource. It gets created, but it's `create` func doesn't wait for it to be "ready", just "exist". 

So when the `AttachVolume` comes along, the volume is still `creating` (or something), so the call simply fails. Here we try to account for that and retry up to 1 minute. 

Here in this block I retry until I get that it's attached or attaching. 

It's kind of an edge case, maybe. Could I clear this up? Unfortunately I ran into this issue early on, and never hit the error case later in development :/

**Comment by julienlavergne** [comment]
Create ebs volume lifecycle is just like any other aws resources. TF just needs to wait for availability of any resources (e.g. aws eip ressource needs an aws instance). 

Exist func must mean the ressource was well created.

File: builtin/providers/aws/resource_aws_volume_attachment.go (hunk 2 above)
Comment (phinze) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: > In TF's case, say that volume is another resource. It gets created, but it's create func doesn't wait for it to be "ready", just "exist". 

Seems like it should be the volume's responsibility to wait to return until the volume is `ready`, such that subsequent resources don't have to worry about it.

So then we can expect a single call to `AttachVolumes` to do its thing (or be able to error right away), then we just poll to wait for its thing to be done.

Let me know if that makes sense.

Comment (julienlavergne) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: @phinze

Currently aws_ebs_volume doesn’t wait if the volume is ready.

That makes sense that resource aws_volume_attachment doesn’t wait for the availability of the volume but that aws_ebs_volume guarantees it by its return.

-- 
Julien Lavergne
Sent with Airmail

On May 27, 2015 at 4:39:32 PM, Paul Hinze (notifications@github.com) wrote:

In builtin/providers/aws/resource_aws_volume_attachment.go:

> +              Delay:      5 * time.Second,
> +          MinTimeout: 3 * time.Second,
> +  }
> +
> +  _, err := stateConf.WaitForState()
> +  if err != nil {
> +          return fmt.Errorf("Error attaching volume %s to instance %s: %s", vID, iID, err)
> +  }
> +
> +  d.SetId(volumeAttachmentID(name, vID, iID))
> +  return resourceAwsVolumeAttachmentRead(d, meta)
> +}
> +
> +func attachVolumeFunc(conn *ec2.EC2, opts *ec2.AttachVolumeInput) resource.StateRefreshFunc {
> +  return func() (interface{}, string, error) {
> +          va, err := conn.AttachVolume(opts)
In TF's case, say that volume is another resource. It gets created, but it's create func doesn't wait for it to be "ready", just "exist".

Seems like it should be the volume's responsibility to wait to return until the volume is ready, such that subsequent resources don't have to worry about it.

So then we can expect a single call to AttachVolumes to do its thing (or be able to error right away), then we just poll to wait for its thing to be done.

Let me know if that makes sense.

Comment (catsby) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: > Let me know if that makes sense.

:+1:

**Comment by catsby** [comment]
@phinze refactored – take another look please

ALSO – I think the acc. test leaks a volume, but I haven't had time to crack down on that.

File: builtin/providers/aws/resource_aws_volume_attachment.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	_, err := conn.AttachVolume(opts)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			return fmt.Errorf("[WARN] Error attaching volume (%s) to instance (%s), message: \"%s\", code: \"%s\"",
+				vID, iID, awsErr.Message(), awsErr.Code())
+		}
+		return err
+	}
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: So now that we loop above, there's no eventual consistency here? No need to loop and wait for "attached" or anything?

**Comment by phinze** [comment]
Clean clean clean! :heart_eyes_cat: 

One question, but if the answer is "nope" this LGTM.

**Comment by julienlavergne** [comment]
no need to wait for attachment; ebs volume needs just to be created above in the platform creation process.

**Comment by julienlavergne** [comment]
LGTM :)

File: builtin/providers/aws/resource_aws_volume_attachment.go (hunk 3 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: added polling for `attached`9329073 , JUST FOR YOU @phinze

**Comment by catsby** [comment]
1f5c038: _now with documentation!_

**Comment by catsby** [comment]
> ALSO – I think the acc. test leaks a volume, but I haven't had time to crack down on that.

I plugged the leak! 

yeah

**Comment by mzupan** [comment]
just used this and it works great

**Comment by phinze** [comment]
:+1: x :100: LGTM!

**Comment by julienlavergne** [comment]
Thx !


Julien

**Comment by catsby** [comment]
Thanks for the validation @mzupan ! Merging...

**Comment by Pryz** [comment]
It seems that the volume is attached after the provisioning of the instance. Is it expected ? If yes, how can we managed the bootstrap of the volume ?

**Comment by julienlavergne** [comment]
It's right and the expected behavior during the instance boot.

Envoyé de mon iPhone

**Comment by Pryz** [comment]
Thanks. Do you have an example on how to manage the bootstrap of the volume ? (mkfs, etc etc)

**Comment by phinze** [comment]
@Pryz one way would be to add a `remote-exec` provisioner to the volume attachment itself and give it connection details for the instance, something like this:

```js
resource "aws_volume_attachment" "ebs_att" {
  device_name = "/dev/sdh"
  volume_id = "${aws_ebs_volume.example.id}"
  instance_id = "${aws_instance.web.id}"
  // this will run after the attachment is made
  provisioner "remote-exec" {
    connection {
      host = "${aws_instance.web.public_ip}"
      // ...
    }
    inline = [ "mkfs -t ext4 /dev/sdh" /* ... */ ]
  }
}
```

**Comment by julienlavergne** [comment]
Cloudinit does the thing very well

-- 
Julien Lavergne
Sent with Airmail

On July 27, 2015 at 4:39:32 PM, Paul Hinze (notifications@github.com) wrote:

@Pryz one way would be to add a remote-exec provisioner to the volume attachment itself and give it connection details for the instance, something like this:

resource "aws_volume_attachment" "ebs_att" {
  device_name = "/dev/sdh"
  volume_id = "${aws_ebs_volume.example.id}"
  instance_id = "${aws_instance.web.id}"
  // this will run after the attachment is made
  provisioner "remote-exec" {
    connection {
      host = "${aws_instance.web.public_ip}"
      // ...
    }
    inline = [ "mkfs -t ext4 /dev/sdh" /* ... */ ]
  }
}

**Comment by Pryz** [comment]
The proposition of @phinze is working. I really think that if we can do that directly with a "ebs_block_device" resource things would be more readable and easier to maintain.

Also the @phinze's proposition involve code duplication (about the connection).

@julienlavergne The issue I have with cloudinit is that I have a lot of different instances to maintain. Most of the time the only differences (from Terraform) are about EBS. So having a "generic" cloud-init/user-data and manage the EBS with ebs_block_device would be easier :)

**Comment by julienlavergne** [comment]
I understand this feature.

-- 
Julien Lavergne
+33670273883
http://julienlavergne.com
Sent with Airmail

On September 4, 2015 at 12:05:26 PM, Julien Fabre (notifications@github.com) wrote:

The proposition of @phinze is working. I really think that if we can do that directly with a "ebs_block_device" resource things would be more readable and easier to maintain.

Also the @phinze's proposition involve code duplication (about the connection).

@julienlavergne The issue I have with cloudinit is that I have a lot of different instances to maintain. Most of the time the only differences (from Terraform) are about EBS. So having a "generic" cloud-init/user-data and manage the EBS with ebs_block_device would be easier :)

**Pull request merged**